README.md
cordova-background-geolocation-firebase &middot; [![npm](https://img.shields.io/npm/dm/cordova-background-geolocation-firebase.svg)]() [![npm](https://img.shields.io/npm/v/cordova-background-geolocation-firebase.svg)]()
============================================================================

[![](https://dl.dropboxusercontent.com/s/nm4s5ltlug63vv8/logo-150-print.png?dl=1)](https://www.transistorsoft.com)

-------------------------------------------------------------------------------

Firebase Proxy for [Cordova Background Geolocation](https://github.com/transistorsoft/cordova-background-geolocation-lt).  The plugin will automatically post locations to your own Firebase database, overriding the `cordova-background-geolocation` plugin's SQLite / HTTP services.

![](https://dl.dropboxusercontent.com/s/2ew8drywpvbdujz/firestore-locations.png?dl=1)

----------------------------------------------------------------------------

The **[Android module](https://shop.transistorsoft.com/collections/frontpage/products/background-geolocation-firebase)** requires [purchasing a license](https://shop.transistorsoft.com/collections/frontpage/products/background-geolocation-firebase).  However, it *will* work for **DEBUG** builds.  It will **not** work with **RELEASE** builds [without purchasing a license](https://shop.transistorsoft.com/collections/frontpage/products/background-geolocation-firebase).

----------------------------------------------------------------------------

# Contents
- ### [Configuration Options](#large_blue_diamond-configuration-options)
- ### [Installing the Plugin](#large_blue_diamond-installing-the-plugin)
- ### [Setup Guide](#large_blue_diamond-setup-guide)
- ### [Using the plugin](#large_blue_diamond-using-the-plugin)
- ### [Example](#large_blue_diamond-example)

## :large_blue_diamond: Installing the Plugin

### From npm

```
$ cordova plugin add cordova-background-geolocation-firebase --save
```

## :large_blue_diamond: Setup Guide

You also need to download **`google-services.json`** on Android and **`GoogleService-Info.plist`** on iOS from **Firebase Console**.  Place them into the cordova project root folder. Then use `<resource-file>` to copy those files into appropriate folders:

```xml
<platform name="android">
    <resource-file src="google-services.json" target="app/google-services.json" />
    ...
</platform>
<platform name="ios">
    <resource-file src="GoogleService-Info.plist" />
    ...
</platform>
```

### Android

- Add the following `<preference>` to your `config.xml`

```xml
<widget>
  .
  .
  .
  <preference name="GradlePluginGoogleServicesEnabled" value="true" />
  <preference name="GradlePluginGoogleServicesVersion" value="4.2.0" />

  <platform name="android">
    <preference name="AndroidXEnabled" value="true" />
    .
    .
    .
  </platform>
</widget>
```

- Add the following plugin to your app (prefix with `ionic` if using Ionic):

```bash
$ cordova plugin add cordova-plugin-androidx-adapter
```

### iOS

This plugin imports iOS `Firebase` dependencies via **Cocoapods**.  If you're unfamiliar with **Cocoapods**, [see here](https://guides.cocoapods.org/using/getting-started.html) to ensure you have it setup on your system.

After installing the plugin and adding the iOS platform, you must execute `$ pod install`:

```
$ cd platforms/ios
$ pod install
```

:warning: If you haven't update your pods in a while:
```
$ pod update
```

:warning: There's a cordova build issue I haven't yet been able to figure out:  `cordova build ios` **will no longer work**:

```
$ cordova build ios

 fatal error:
      'RxLibrary/GRXWriteable.h' file not found
#import <RxLibrary/GRXWriteable.h>
        ^~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

As a result, you'll have to build your app using XCode from **`YourProject.xcworkspace`** (**not** `YourProject.xcodeproj`).

### Configure your license

:information_source: If you don't *have* a license yet, the plugin is fully functional in `DEBUG` builds so you can try it before [purchasing a license](https://shop.transistorsoft.com/collections/frontpage/products/background-geolocation-firebase).

1. Login to Customer Dashboard to generate an application key:
[www.transistorsoft.com/shop/customers](http://www.transistorsoft.com/shop/customers)
![](https://gallery.mailchimp.com/e932ea68a1cb31b9ce2608656/images/b2696718-a77e-4f50-96a8-0b61d8019bac.png)

2. Add your license-key:

```
$ cordova plugin remove cordova-background-geolocation-firebase
$ cordova plugin add cordova-background-geolocation-firebase --variable LICENSE=YOUR_KEY_HERE
```

## :large_blue_diamond: Using the plugin ##

The plugin creates the object **`window.BackgroundGeolocationFirebase`**.

```javascript
var bgGeoFirebase = window.BackgroundGeolocationFirebase;
```

### Ionic 2+:

```javascript
let bgGeoFirebase = (<any>window).BackgroundGeolocationFirebase;
```

## :large_blue_diamond: Example

```javascript

onDeviceReady() {
  var bgGeoFirebase = window.BackgroundGeolocationFirebase;
  bgGeoFirebase.configure({
    locationsCollection: 'locations',
    geofencesCollection: 'geofences'
  });

  // Now configure background-geolocation as normal.  Nothing changes here.
  var bgGeo = window.BackgroundGeolocation;
  bgGeo.on('location', function(location) {...});
  bgGeo.ready({
    .
    .
    .
  }, function(state) {
    console.log('[BackgroundGeolocation] ready -', state);
  });
}

```

### Ionic 2+

```javascript
import { Component } from '@angular/core';
import { NavController, Platform } from 'ionic-angular';

@Component({
  selector: 'page-home',
  templateUrl: 'home.html'
})
export class HomePage {

  constructor(public navCtrl: NavController, public platform: Platform) {

  }

  ionViewDidLoad() {
    this.platform.ready().then(() => {
      this.configureBackgroundGeolocation();
    });
  }

  configureBackgroundGeolocation() {
    let bgGeoFirebase = (<any>window).BackgroundGeolocationFirebase;

    bgGeoFirebase.configure({
      locationsCollection: 'locations',
      geofencesCollection: 'geofences'
    });

    let bgGeo = (<any>window).BackgroundGeolocation;

    bgGeo.on('location', (location) => {
      console.log('[location] - ', location);
    });

    bgGeo.ready({
      stopOnTerminate: false,
      debug: true
    }, (state) => {
      if (!state.enabled) {
        bgGeo.start();
      }
    });
  }
}


```

## :large_blue_diamond: Firebase Functions

`BackgroundGeolocation` will post its default "Location Data Schema" to your Firebase app.

```json
{
  "location":{},
  "param1": "param 1",
  "param2": "param 2"
}
```

You should implement your own [Firebase Functions](https://firebase.google.com/docs/functions/) to "*massage*" the incoming data in your collection as desired.  For example:

```typescript
import * as functions from 'firebase-functions';

exports.createLocation = functions.firestore
  .document('locations/{locationId}')
  .onCreate((snap, context) => {
    const record = snap.data();

    const location = record.location;

    console.log('[data] - ', record);

    return snap.ref.set({
      uuid: location.uuid,
      timestamp: location.timestamp,
      is_moving: location.is_moving,
      latitude: location.coords.latitude,
      longitude: location.coords.longitude,
      speed: location.coords.speed,
      heading: location.coords.heading,
      altitude: location.coords.altitude,
      event: location.event,
      battery_is_charging: location.battery.is_charging,
      battery_level: location.battery.level,
      activity_type: location.activity.type,
      activity_confidence: location.activity.confidence,
      extras: location.extras
    });
});


exports.createGeofence = functions.firestore
  .document('geofences/{geofenceId}')
  .onCreate((snap, context) => {
    const record = snap.data();

    const location = record.location;

    console.log('[data] - ', record);

    return snap.ref.set({
      uuid: location.uuid,
      identifier: location.geofence.identifier,
      action: location.geofence.action,
      timestamp: location.timestamp,
      latitude: location.coords.latitude,
      longitude: location.coords.longitude,
      extras: location.extras,
    });
});

```

## :large_blue_diamond: Configuration Options

#### `@param {String} locationsCollection [locations]`

The collection name to post `location` events to.  Eg:

```javascript
BackgroundGeolocationFirebase.configure({
  locationsCollection: '/locations'
});

BackgroundGeolocationFirebase.configure({
  locationsCollection: '/users/123/locations'
});

BackgroundGeolocationFirebase.configure({
  locationsCollection: '/users/123/routes/456/locations'
});

```

#### `@param {String} geofencesCollection [geofences]`

The collection name to post `geofence` events to.  Eg:

```javascript
BackgroundGeolocationFirebase.configure({
  geofencesCollection: '/geofences'
});

BackgroundGeolocationFirebase.configure({
  locationsCollection: '/users/123/geofences'
});

BackgroundGeolocationFirebase.configure({
  locationsCollection: '/users/123/routes/456/geofences'
});

```


#### `@param {Boolean} updateSingleDocument [false]`

If you prefer, you can instruct the plugin to update a *single document* in Firebase rather than creating a new document for *each* `location` / `geofence`.  In this case, you would presumably implement a *Firebase Function* to deal with updates upon this single document and store the location in some other collection as desired.  If this is your use-case, you'll also need to ensure you configure your `locationsCollection` / `geofencesCollection` accordingly with an even number of "parts", taking the form `/collection_name/document_id`, eg:

```javascript
BackgroundGeolocationFirebase.configure({
  locationsCollection: '/locations/latest'  // <-- 2 "parts":  even
});

// or
BackgroundGeolocationFirebase.configure({
  locationsCollection: '/users/123/routes/456/the_location'  // <-- 4 "parts":  even
});

// Don't use an odd number of "parts"
BackgroundGeolocationFirebase.configure({
  locationsCollection: '/users/123/latest_location'  // <-- 3 "parts": odd!!  No!
});

```


# License

The MIT License (MIT)

Copyright (c) 2018 Chris Scott, Transistor Software

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.




CHANGELOG.md
# CHANGELOG

## [0.3.0] -- 2020-11-02
* [Changed] Updated setup for AndroidX, `cordova-android >= 10` now natively supports preferences `GradlePluginGoogleServicesEnabled`, `GradlePluginGoogleServicesVersion`.

## [0.2.1] -- 2020-09-24
* [Fixed][Android] Incorrect path to tslocationmanager.aar after upgrade to gradle:4.0.0.

## [0.2.0] -- 2020-07-13
* [Fixed][Android] `com.android.tools.build:gradle:4.0.0` no longer allows "*direct local aar dependencies*".

## [0.1.1] -- 2019-09-11
- [iOS] Fix issue #11:  Provide defaults for `locationsCollection`, `geofencesCollection`
- [Added] Capacitor support.
- [Changed] Re-organize android `tsfirebaseproxy.aar` library location to simply `/libs/tsfirebaseproxy`.

## [0.1.0] -- 2019-06-04
- [Android] Fixed bug with `geofencesCollection`
- [Android] Update Firebase dependencies
- [Android] Update `cordova-support-google-services: ~1.3.1`


www/BackgroundGeolocationFirebase.js
/***
 * Custom Cordova Background Fetch plugin. 
 * @author <chris@transistorsoft.com>
 * iOS native-side is largely based upon http://www.mindsizzlers.com/2011/07/ios-background-location/
 * 
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
*/
var exec = require("cordova/exec");
var MODULE = "BackgroundGeolocationFirebase";
module.exports = {
    
    configure: function(config, success, failure) {
        config = config || {};
        success = success || function() {};
        failure = failure || function() {};
        exec(success, failure, MODULE, 'configure', [config]);
    }    
};


src/android/CDVBackgroundGeolocationFirebase.java
package com.transistorsoft.cordova.bggeo;

import com.transistorsoft.locationmanager.logger.TSLog;
import com.transistorsoft.tsfirebaseproxy.TSFirebaseProxy;

import org.json.JSONArray;
import org.json.JSONObject;
import org.json.JSONException;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;

import android.app.Activity;

public class CDVBackgroundGeolocationFirebase extends CordovaPlugin {
    public static final String TAG = "TSLocationManager";
    public static final String ACTION_CONFIGURE = "configure";

    private boolean isRegistered;

    @Override
    protected void pluginInitialize() {
        isRegistered = false;
    }

    public boolean execute(String action, JSONArray data, CallbackContext callbackContext) throws JSONException {
        boolean result = false;

        TSLog.logger.info(TSLog.info("[Firebase] $ " + action));

        if (ACTION_CONFIGURE.equalsIgnoreCase(action)) {
            result = true;
            configure(data.getJSONObject(0), callbackContext);
        }
        return result;
    }

    private void configure(JSONObject options, final CallbackContext callbackContext) throws JSONException {
        TSLog.logger.info(TSLog.info("[Firebase] - configure: " + options));
        Activity activity = cordova.getActivity();

        TSFirebaseProxy proxy = TSFirebaseProxy.getInstance(activity.getApplicationContext());

        if (options.has(TSFirebaseProxy.FIELD_LOCATIONS_COLLECTION)) {
            proxy.setLocationsCollection(options.getString(TSFirebaseProxy.FIELD_LOCATIONS_COLLECTION));
        }
        if (options.has(TSFirebaseProxy.FIELD_GEOFENCES_COLLECTION)) {
            proxy.setGeofencesCollection(options.getString(TSFirebaseProxy.FIELD_GEOFENCES_COLLECTION));
        }
        if (options.has(TSFirebaseProxy.FIELD_UPDATE_SINGLE_DOCUMENT)) {
            proxy.setUpdateSingleDocument(options.getBoolean(TSFirebaseProxy.FIELD_UPDATE_SINGLE_DOCUMENT));
        }
        proxy.save(activity.getApplicationContext());

        if (!isRegistered) {
            isRegistered = true;
            proxy.register(activity.getApplicationContext());
        }
        callbackContext.success();
    }
}


src/ios/CDVBackgroundGeolocationFirebase.m
////
//  CDVBackgroundGeoLocation
//
//  Created by Chris Scott <chris@transistorsoft.com> on 2013-06-15
//  Largely based upon http://www.mindsizzlers.com/2011/07/ios-background-location/
//
#import "CDVBackgroundGeolocationFirebase.h"

#import "Firebase.h"
#import <TSLocationManager/TSLocationManager.h>

static NSString *const TAG = @"CDVBackgroundGeolocationFirebase";

static NSString *const PERSIST_EVENT                = @"TSLocationManager:PersistEvent";

static NSString *const FIELD_LOCATIONS_COLLECTION = @"locationsCollection";
static NSString *const FIELD_GEOFENCES_COLLECTION = @"geofencesCollection";
static NSString *const FIELD_UPDATE_SINGLE_DOCUMENT = @"updateSingleDocument";

static NSString *const DEFAULT_LOCATIONS_COLLECTION = @"locations";
static NSString *const DEFAULT_GEOFENCES_COLLECTION = @"geofences";


@implementation CDVBackgroundGeolocationFirebase
{
    BOOL isRegistered;
}

- (void)pluginInitialize
{
    isRegistered = NO;
    _locationsCollection = DEFAULT_LOCATIONS_COLLECTION;
    _geofencesCollection = DEFAULT_GEOFENCES_COLLECTION;
    _updateSingleDocument = NO;

    if (![FIRApp defaultApp]) {
        [FIRApp configure];
    }
}

- (void) configure:(CDVInvokedUrlCommand*)command
{
    NSDictionary *config = [command.arguments objectAtIndex:0];

    if (config[FIELD_LOCATIONS_COLLECTION]) {
        _locationsCollection = config[FIELD_LOCATIONS_COLLECTION];
    }
    if (config[FIELD_GEOFENCES_COLLECTION]) {
        _geofencesCollection = config[FIELD_GEOFENCES_COLLECTION];
    }
    if (config[FIELD_UPDATE_SINGLE_DOCUMENT]) {
        _updateSingleDocument = [config[FIELD_UPDATE_SINGLE_DOCUMENT] boolValue];
    }

    if (!isRegistered) {
        TSConfig *bgGeo = [TSConfig sharedInstance];
        [bgGeo registerPlugin:@"firebaseproxy"];
        isRegistered = YES;
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(onPersist:)
                                                     name:PERSIST_EVENT
                                                   object:nil];
    }
}

-(void) onPersist:(NSNotification*)notification {
    NSDictionary *data = notification.object;
    NSString *collectionName = (data[@"location"][@"geofence"]) ? _geofencesCollection : _locationsCollection;

    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
        FIRFirestore *db = [FIRFirestore firestore];
        // Add a new document with a generated ID
        if (!self.updateSingleDocument) {
            __block FIRDocumentReference *ref = [[db collectionWithPath:collectionName] addDocumentWithData:notification.object completion:^(NSError * _Nullable error) {
                if (error != nil) {
                    NSLog(@"Error adding document: %@", error);
                } else {
                    NSLog(@"Document added with ID: %@", ref.documentID);
                }
            }];
        } else {
            [[db documentWithPath:collectionName] setData:notification.object completion:^(NSError * _Nullable error) {
                if (error != nil) {
                    NSLog(@"Error writing document: %@", error);
                } else {
                    NSLog(@"Document successfully written");
                }
            }];
        }
    });
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

@end

